### ISSUE TITLE
LSP4J 0.13.0 

### ISSUE THREAD

**Issue by jonahgraham** [eclipse/lsp4j#541]
This is the Release plan and TODO list for LSP4J release v0.13.0.

## Steps for Release

Items at the beginning of development

- [x] Create an Endgame Issue to track the release. As a starting point use [documentation/releasing.md](https://github.com/eclipse/lsp4j/blob/master/documentation/releasing.md).
    - [x] Add the [Endgame](https://github.com/eclipse/lsp4j/labels/endgame) label and Milestone for the release
    - [x] Make sure any previous edits made to [Endgame issues](https://github.com/eclipse/lsp4j/labels/endgame) of previous releases are updated in [documentation/releasing.md](https://github.com/eclipse/lsp4j/blob/master/documentation/releasing.md)
- [x] Ensure all previous [Endgame issues](https://github.com/eclipse/lsp4j/labels/endgame) are done.
- [x] Create a [New milestone](https://github.com/eclipse/lsp4j/milestones/new) for the release
- [x] Create release on [PMI](https://projects.eclipse.org/projects/technology.lsp4j)
- [x] Check [CHANGELOG.md](https://github.com/eclipse/lsp4j/blob/master/CHANGELOG.md) is up to date. The changelog should have a version entry, release date, API Breakages and other information consistent with current entries in the changelog.
- [x] Check [README.md](https://github.com/eclipse/lsp4j/blob/master/README.md) is up to date. In particular that the planned release and which versions of DAP and LSP are support is listed.
- [x] Increment version of all feature.xml, pom.xml and any other place full version is used. (Easiest way is global find and replace, e.g. `s/0.12.0/0.13.0/g` and review changes.) Ensure that `-SNAPSHOT` is restored in the [gradle/versions.gradle](https://github.com/eclipse/lsp4j/blob/master/gradle/versions.gradle) and  [releng/pom.xml](https://github.com/eclipse/lsp4j/blob/master/releng/pom.xml)
- [x] Enable `sh './releng/deploy-build.sh'` in [releng/build.Jenkinsfile](https://github.com/eclipse/lsp4j/blob/master/releng/build.Jenkinsfile) 
- [x] Ensure [the CI build](https://ci.eclipse.org/lsp4j/job/lsp4j-multi-build/job/main/) is stable - it is always better to release a "Green Dot" build

Items in the days ahead of Release day:

- [x] Schedule the release and if needed schedule a release review on the [PMI](https://projects.eclipse.org/projects/technology.lsp4j). A release review is needed every 12 months, not with each release.
- [x] Check [CHANGELOG.md](https://github.com/eclipse/lsp4j/blob/master/CHANGELOG.md) is up to date. The changelog should have a version entry, release date, API Breakages and other information consistent with current entries in the changelog.
- [x] Check [README.md](https://github.com/eclipse/lsp4j/blob/master/README.md) is up to date. In particular that the planned release and which versions of DAP and LSP are support is listed.
- [x] Check all closed PRs and Issues to make sure their milestone is set. (*Note:* this was not until after 0.10.0 release so many old PRs and Issues have no milestone, therefore only consider items back to approx 5 Nov 2020). [This search may be useful to identify such closed issues](https://github.com/eclipse/lsp4j/issues?q=is%3Aclosed+no%3Amilestone+updated%3A%3E%3D2020-11-06)
- [x] Create and analyse a [`japicmp`](https://siom79.github.io/japicmp/) report and publish it as part of the build. Ensure that the API versions are incremented accurately based on the report.

Items on Release day:

- [x] Prepare the repo for release by:
    - [x] removing `-SNAPSHOT` from [gradle/versions.gradle](https://github.com/eclipse/lsp4j/blob/master/gradle/versions.gradle)
    - [x] removing `-SNAPSHOT` from [releng/pom.xml](https://github.com/eclipse/lsp4j/blob/master/releng/pom.xml) entries in `<dependencies>` section.
    - [x] disabling `sh './releng/deploy-build.sh'` in [releng/build.Jenkinsfile](https://github.com/eclipse/lsp4j/blob/master/releng/build.Jenkinsfile) 
    - see commit https://github.com/eclipse/lsp4j/commit/328ce8a4c89b0cd84fb62118f459b6cf79b09e90 for a past example
- [x] Push the above change
- [x] Run [the CI build](https://ci.eclipse.org/lsp4j/job/lsp4j-multi-build/job/master/)
- [x] Mark the build as Keep Forever and add to the description `v0.13.0`
- [x] Deploy the release by running [the Release CI job](https://ci.eclipse.org/lsp4j/job/lsp4j-release-eclipse) with parameters:
    - `LSP4J_PUBLISH_LOCATION` -> `updates/releases/0.13.0` ( <-- check version number)
    - `PROJECT` -> `lsp4j-multi-build/job/master`
    - `LSP4J_BUILD_NUMBER` -> the build that was just run above
- [x] Add to the deploy job description `v0.13.0`
- [x] Promote the staged repository to maven central
    - [Login to Nexus](https://oss.sonatype.org/#stagingRepositories)
        - To obtain permission add request to [OSSRH-26079](https://issues.sonatype.org/browse/OSSRH-26079)
    - go to *Staging Repositories*, after a short delay the staged LSP4J release should appear
    - click the staged LSP4J repo
    - press the *Close* button located in the toolbar. This runs activities, including checking various rules
    - once the rules are done (if successful), press the *Release* button (you may need to press *Refresh* to enable the *Release* button)
    - check https://search.maven.org/search?q=g:org.eclipse.lsp4j to make sure the latest release has arrived - this takes a while, 15 minutes for the files to be [on the server](https://repo1.maven.org/maven2/org/eclipse/lsp4j/) and even longer for the [search indexes](https://search.maven.org/search?q=g:org.eclipse.lsp4j) to update
- [x] Update the meta-data on [PMI downloads page](https://projects.eclipse.org/projects/technology.lsp4j/downloads)
- [x] Tag the release. Example: `git tag -a v0.13.0 HEAD -m"LSP4J 0.13.0" && git push origin v0.13.0`
- [x] Contribute to Simrel. See [Simrel contribution example](https://git.eclipse.org/r/#/c/158624/)
- [x] Create a [release page on github](https://github.com/eclipse/lsp4j/releases/new)
- [x] Make an announcement on lsp4j-dev based on the [release page on github](https://github.com/eclipse/lsp4j/releases/tag/v0.13.0). [Example on lsp4j-dev archives](https://www.eclipse.org/lists/lsp4j-dev/msg00063.html)
- [x] Update [documentation/releasing.md](https://github.com/eclipse/lsp4j/blob/master/documentation/releasing.md) with any changes that may have been made to the release process.
- [x] Create the endgame for the next release right away, especially as version numbers and restoring `-SNAPSHOT` need to be done right away.

**Comment by jonahgraham** [comment]
Following discussions in #542 I have added as a step to run [`japicmp`](https://siom79.github.io/japicmp/). For this first release to use it I don't know exactly what that will mean.

**Comment by tjpalmer** [comment]
Is there an eta on this?

**Comment by jonahgraham** [comment]
@tjpalmer - no eta - it is a somewhat reactive process if there is demand for it or a critical fix we need to roll out. What are your plans? Are you depending on a not yet released part of the code?

**Comment by tjpalmer** [comment]
Thanks much for the quick reply and for your work on this! I was curious about the `allowedTags` for markdown, and I saw that was on main but not in the current release. It's not an urgent thing, but that's why I asked. Maybe other things will come up later.

**Comment by jonahgraham** [comment]
Thank you for the specific feature you mentioned (`allowedTags`) I had a look around and I think that there are starting to accumulate some changes. We should look at doing a release once 3.17.0 spec is finalized. In the meantime we publish snapshots so you can try out 3.17.0 version of the spec in progress. Does that work for you? Let me (us) know if there is anything else you need in the meantime.

**Comment by jonahgraham** [comment]
There is starting to be a lot of unreleased stuff in LSP4J, should we continue to wait on LSP 3.17.0 being finalized, or do like we did in LSP4J 0.10 and release a _(specification is not finalized yet)_ version?

Thanks to @KamasamaK in PR #605 LSP4J now has all the DAP API up until 1.55.

**Comment by KamasamaK** [comment]
If anyone is curious how long of a wait that might be, the primary maintainer said on March 1st that "The plan is to have a new 3.17 release the next ~ 2 month".

https://github.com/microsoft/language-server-protocol/issues/1344#issuecomment-1055375678

**Comment by cdietrich** [comment]
3.17 was published today.
would be nice to have a release soon so that downstream projects still have a bit of time to adapt

**Comment by jonahgraham** [comment]
👍 OK @cdietrich lets schedule a release. I can do the work probably 19/23/24 of this month. How does release date of 24th sound? If you think it would matter to have it earlier, please propose a date and I will figure out if I can do that (or someone else can do the releng 😉). 

There are still some outstanding tasks for the release - https://github.com/eclipse/lsp4j/milestone/20 - however we could just not do that Java 11 one yet, as it has never been important enough to me for me to spend time on up until now.

@KamasamaK / @cdietrich  and anyone else reading this, if there are other issues that should be in 0.13.0 please update their milestone.

**Comment by jonahgraham** [comment]
I chose 24th because that was in the m3 week. When do you recommend?

**Comment by cdietrich** [comment]
the friday before so that i can push the xtext patch for that in time
@mickaelistria do you have already plans for lsp4e adaptations

**Comment by cdietrich** [comment]
thursday would also be ok for me.

**Comment by cdietrich** [comment]
@KamasamaK you seem to do a diff if we miss something in spec.
is this already done with the last merge or do you need to do another round?

**Comment by KamasamaK** [comment]
@cdietrich I believe the only thing we are missing from the 3.17 spec is Notebook support. That's a big one that I still need to look into. I am in the process of putting together a PR now with another set of various changes, such as pulling in more items into `Protocol.xtend`. Other than that, I just need to go through and remove all of the pertinent `@Beta` annotations before the release.

**Comment by jonahgraham** [comment]
I have done the PMI stuff and requested a release review (it has been > 1 year since our last one): https://projects.eclipse.org/projects/technology.lsp4j/reviews/0.13.0-release-review 

I need to do the IP log, which may mean getting dash license running on LSP4J. That needs to be submitted ASAP so we can meet the release day timeline.

I also sent an intention email to [lsp4j ](https://www.eclipse.org/lists/lsp4j-dev/msg00105.html)and [lsp4e](https://www.eclipse.org/lists/lsp4e-dev/msg00665.html) (same email)

**Comment by jonahgraham** [comment]
> 3.17 spec is Notebook support

Can you update readme and changelog to say that please if it isn't likely to be done for this release?

**Comment by jonahgraham** [comment]
https://gitlab.eclipse.org/eclipsefdn/emo-team/emo/-/issues/290 is the EMO tracking issue for the release

**Comment by cdietrich** [comment]
Did we have any dep changes besides minor gson bump

**Comment by cdietrich** [comment]
So I assume no dash needed

**Comment by jonahgraham** [comment]
I think we are pretty much ready to go. Just a few items with 0.13.0 milestone: https://github.com/eclipse/lsp4j/milestone/20

> Create and analyse a [japicmp](https://siom79.github.io/japicmp/) report and publish it as part of the build. Ensure that the API versions are incremented accurately based on the report.

See #626 for status of this item.

**Comment by jonahgraham** [comment]
We have our [release review approved](https://gitlab.eclipse.org/eclipsefdn/emo-team/emo/-/issues/290#note_755534) and apart from the japicmp (#626) which I will finish off today we are good to go.

Therefore I will be doing the "Items on Release day:" part of the checklist over the course of the day today.

**Comment by jonahgraham** [comment]
And we are done here! https://github.com/eclipse/lsp4j/releases/tag/v0.13.0

On to the next one: https://github.com/eclipse/lsp4j/issues/628

**Comment by cdietrich** [comment]
thx @jonahgraham
